### PR TITLE
Feature/73 support hybrid login

### LINKED ISSUE #72: Read and write to new cookie
Currently, tokens are read and written from cookies with the key `SC_TOKEN`. We should introduce a new cookie key for tokens from the new service, so they don't clash with existing tokens and so users can be authenticated against old and new services at the same time.

### LINKED ISSUE #73: Support Hybrid Login
- Integrate with new token service
- Read & write new token to new cookie
- Send new flag to login page

### PR DESCRIPTION
**Pull request by cjhowald**
Fixes #73, Fixes #72

### CODE REVIEW

**Comment by dami-gg** [comment]
Please don't include the distribution files

File: src/authentication-service.js [hunk 1]
```diff
@@ -159,26 +169,43 @@
...
+          return $q.reject(response.reason);
+        });
+    }
+
+    function doLogin(email, password, useOldToken) {
+      return $injector.get('$http').post(ScEnvironmentsService.getTokensAPI(self.getEnvironment(), useOldToken),
         {
           email: email,
-          password: password,
-          agent: getUserAgent()
```
Comment (dami-gg) [inline src/authentication-service.js]: Thanks for removing that 👍

File: src/authentication-service.js [hunk 2]
```diff
@@ -402,8 +424,8 @@
     /**
      * Removes the token from the cookie.
      */
-    function clearToken() {
-      removeCookies(TOKEN_COOKIE_KEY);
+    function clearToken(useOldToken) {
+      removeCookies(getTokenKey(useOldToken));
```
Comment (dami-gg) [inline src/authentication-service.js]: I believe here we should delete both token cookies to keep consistency

File: src/environments-service.js [hunk 3]
```diff
@@ -35,8 +35,12 @@
       return '/logout';
     }
 
-    function getTokensAPI(environment) {
-      return environment.TOKEN_SERVICE + '/tokens';
+    function getTokensAPI(environment, useOldToken) {
+      if (useOldToken) {
+        return environment.TOKEN_SERVICE + '/tokens'; // TODO change to old key
+      } else {
+        return environment.USER_SERVICE_NEW + '/tokens'; // TODO change to new key
```
Comment (dami-gg) [inline src/environments-service.js]: I understand that we need to update and release the communicator before releasing this right?

File: src/authentication-service.js [hunk 4]
```diff
@@ -159,26 +169,43 @@
...
      * @param password
+     * @param expectOldToken service requesting auth is expecting an old token
      * @returns {Function} A resolved promise with the token in case the user could be logged in or a rejected one in any other case
      */
-    function login(email, password) {
-      return $injector.get('$http').post(ScEnvironmentsService.getTokensAPI(self.getEnvironment()),
+    function login(email, password, expectOldToken) {
+      return $q.allSettled([doLogin(email, password, true), doLogin(email, password, false)])
+        .then(function(responses) {
+          var response = expectOldToken && responses[0] || responses[1];
```
Comment (dami-gg) [inline src/authentication-service.js]: just be aware that this code is not equivalent to: expectOldToken ? responses[0] : responses[1] which is what I think you wanted to use

File: src/authentication-service.js [hunk 5]
```diff
@@ -141,10 +150,11 @@
      * Redirects to home page in case there is a user already authenticated
      * To be used in the login route's resolve method to prevent users seeing the login dialog when they are already
      * authenticated
+     * @param useOldToken use old token for authentication
      * @returns {*} A promise since it's the value expected by the resolve method
      */
-    function redirectToHomeIfAuthenticated() {
-      if (service.isAuthenticated()) {
+    function redirectToHomeIfAuthenticated(useOldToken) {
```
Comment (dami-gg) [inline src/authentication-service.js]: Here and in several more methods I see that you use a parameter to decide whether to use the old token or not, but since it's a configuration parameter which has to be configured when setting up the login, I think you should directly use the global variable everywhere so nobody is able to alter the behaviour passing a wrong value by mistake.

File: src/authentication-service.js [hunk 6]
```diff
@@ -97,12 +105,13 @@
...
 
-  function ScAuthenticationFactory($window, $location, $q, $localStorage, $cookies, $injector, $timeout, ScEnvironmentsService, jwtHelper) {
+  function ScAuthenticationFactory($window, $location, $q, $localStorage, $cookies, $injector, $timeout, ScEnvironmentsService) {
     'use strict';
 
     var self = this;
 
-    var TOKEN_COOKIE_KEY = "SC_TOKEN";
+    var TOKEN_COOKIE_KEY_OLD = "SC_TOKEN";
+    var TOKEN_COOKIE_KEY = "SC_TOKEN_2";
```
Comment (cjhowald) [inline src/authentication-service.js]: We will need to update login to handle the session + access tokens based on the token migration. But this is currently in design and I think we should make these sweeping changes in another PR.

Comment (cjhowald) [inline src/authentication-service.js]: I created [an issue](https://github.com/zalando-incubator/solution-center-login/issues/78) for this.

File: src/authentication-service.js (hunk 5 above)
Comment (cjhowald) [inline src/authentication-service.js]: There are two things that determine which token to use, configuration and a flag from login. The first is set in the provider and read with calls to `self.shouldUseOldToken()`. I used to read this once and set it to a variable used throughout (as you suggested), but it was confusing when combined the second.

The second configuration is based on what the redirection application wants (e.g. a migrated app needs a new token to authenticate). The solution center parses the `newToken` route param and passes in this flag in `login` and `redirectToHomeIfAuthenticated`. This param is necessary to allow the hybrid login to authenticate old and new applications. In this case, the param is passed to lower-level functions to ensure they use the right token.

Also note that login always authenticates with old and new (via a param passed to `doLogin`) according to the design.

So it is not possible to have a single global variable that controls everything. If there are particular areas that you think can be improved, please note them.

File: src/authentication-service.js (hunk 4 above)
Comment (cjhowald) [inline src/authentication-service.js]: Good catch, fixed.

Now, in the scenario you mentioned, the user will see invalid credentials (since an old token could not be obtained) even though a new token as obtained. This is the correct behavior.

File: src/authentication-service.js (hunk 2 above)
Comment (cjhowald) [inline src/authentication-service.js]: I have changed it so both tokens are cleared on logout. Note that this change is in `clearCredentials`. We can't make the change in `clearToken` itself, since we call `clearToken` in `setToken` before we set a new token. If we clear both when we set one, we will end up in a last-one-wins situation with the tokens. Instead, I call `clearToken` with both false and true from `clearCredentials`.

File: src/environments-service.js (hunk 3 above)
Comment (cjhowald) [inline src/environments-service.js]: Correct. I've created #79 to track this issue.

File: src/authentication-service.js [hunk 7]
```diff
@@ -240,7 +240,8 @@
           // TODO Log error
         })
         .finally(function () {
-          service.clearCredentials();
+          service.clearCredentials(true);
```
Comment (dami-gg) [inline src/authentication-service.js]: Thanks god this is temporal, otherwise no one not involved in this project would understand these two consecutive calls with true and false on the first sight :D

Comment (cjhowald) [inline src/authentication-service.js]: Fair point.

File: src/authentication-service.js (hunk 2 above)
Comment (cjhowald) [inline src/authentication-service.js]: I actually had to modify my original approach, since always clearing both in `clearTokens` was causing the user to get logged out when switching between applications using different services. For example, login to a new service with a new user, get a new token, then attempt to log in to an old service and lose your new token (due to `clearCredentialsAndRedirectToLogin`). Now, we only clear both tokens in the logout scenario.

File: src/authentication-service.js (hunk 5 above)
Comment (dami-gg) [inline src/authentication-service.js]: ok, since this is temporal code it can stay like this I guess

File: src/authentication-service.js (hunk 2 above)
Comment (dami-gg) [inline src/authentication-service.js]: ah yeah, that makes sense Good! :+1:

**Comment by dami-gg** [comment]
:+1:

**Comment by cjhowald** [comment]
👍
